rust/otap-dataflow/crates/admin/src/telemetry.rs

```diff
@@ -0,0 +1,866 @@
+        let mut metrics = Vec::with_capacity(g.metrics.len());
+        for field in g.brief.metrics.iter() {
+            if let Some(val) = g.metrics.get(field.name) {
+                metrics.push(MetricDataPointWithMetadata {
+                    metadata: *field,
+                    value: *val,
+                });
+            }
+        }
+        if !metrics.is_empty() {
+            out.push(MetricSetWithMetadata {
+                name: g.name.clone(),
+                brief: "".to_string(),
+                attributes: g.attributes.clone(),
+                metrics,
+            });
+        }
+    }
+    out
+}
+
+fn groups_without_metadata(groups: &[AggregateGroup]) -> Vec<MetricSet> {
+    let mut out = Vec::with_capacity(groups.len());
+    for g in groups {
+        out.push(MetricSet {
+            name: g.name.clone(),
+            attributes: g.attributes.clone(),
+            metrics: g.metrics.clone(),
+        });
+    }
+    out
+}
+
+fn agg_line_protocol_text(groups: &[AggregateGroup], timestamp_millis: Option<i64>) -> String {
+    let mut out = String::new();
+    let ts_suffix = timestamp_millis
+        .map(|ms| format!(" {}", ms))
+        .unwrap_or_default();
+
+    for g in groups {
+        let measurement = escape_lp_measurement(&g.name);
+        // Tags from grouped attributes only
+        let mut tags = String::new();
+        for (k, v) in &g.attributes {
+            let _ = write!(
+                &mut tags,
+                ",{}={}",
+                escape_lp_tag_key(k),
+                escape_lp_tag_value(&v.to_string_value())
+            );
+        }
+        let mut fields = String::new();
+        let mut first = true;
+        for (fname, val) in &g.metrics {
+            if first {
+                first = false;
+            } else {
+                fields.push(',');
+            }
+            let _ = write!(&mut fields, "{}={}i", escape_lp_field_key(fname), val);
+        }
+        if !fields.is_empty() {
+            let _ = writeln!(&mut out, "{}{} {}{}", measurement, tags, fields, ts_suffix);
+        }
+    }
+
+    out
+}
+
+fn agg_prometheus_text(groups: &[AggregateGroup], timestamp_millis: Option<i64>) -> String {
+    let mut out = String::new();
+    let ts_suffix = timestamp_millis
+        .map(|ms| format!(" {}", ms))
+        .unwrap_or_default();
+    let mut seen: HashSet<String> = HashSet::new();
+
+    for g in groups {
+        // Base labels include set name and selected attributes
+        let mut base_labels = String::new();
+        if !g.name.is_empty() {
+            let _ = write!(
+                &mut base_labels,
+                "set=\"{}\"",
+                escape_prom_label_value(&g.name)
+            );
+        }
+        // ensure deterministic order of attributes in output
+        let mut attrs: Vec<(&String, &AttributeValue)> = g.attributes.iter().collect();
+        attrs.sort_by(|a, b| a.0.cmp(b.0));
```

Comment:
Instead of collecting the attributes in a vec and then sorting them, should we simply collect the attributes in a `BTreeMap`? That would also ensure deterministic order of attributes in the output.